contracts/src/Tracker.sol

```diff
@@ -16,103 +17,109 @@ struct Challenge {
-        IERC20(underlyingToken).safeTransferFrom(msg.sender, address(this), challenges[arxAddress].perUserStake);
-        hasJoined[arxAddress][msg.sender] = true;
-        userChallenges[msg.sender].push(arxAddress);
-        participants[arxAddress].push(msg.sender);
-        challenges[arxAddress].totalStake += challenges[arxAddress].perUserStake;
-        emit Join(msg.sender, arxAddress);
+    function join(uint256 challengeId) public payable {
+        require(challenges[challengeId].startTimestamp != 0, "Challenge does not exist");
+        //require(block.timestamp < challenges[challengeId].startTimestamp, "Challenge has started");
+        IERC20(underlyingToken).safeTransferFrom(msg.sender, address(this), challenges[challengeId].perUserStake);
+        hasJoined[challengeId][msg.sender] = true;
+        userChallenges[msg.sender].push(challengeId);
+        participants[challengeId].push(msg.sender);
+        challenges[challengeId].totalStake += challenges[challengeId].perUserStake;
+        emit Join(msg.sender, challengeId);
     }
 
-    /* Todo: implement timestamp and sender address checks in the signature */
-    function checkIn(address arxAddress, bytes32 msgHash, uint8 v, bytes32 r, bytes32 s) public {
-        uint256 timestamp = block.timestamp;
+    function checkIn(uint256 challengeId, uint256 timestamp, uint8 v, bytes32 r, bytes32 s) public {
         /*require(
-            timestamp <= challenges[arxAddress].endTimestamp && timestamp >= challenges[arxAddress].startTimestamp,
+            timestamp <= challenges[challengeId].endTimestamp && timestamp >= challenges[challengeId].startTimestamp,
             "Invalid timestamp"
         );*/
-        require(hasJoined[arxAddress][msg.sender], "User has not joined the challenge");
-        address recoveredAddr = ECDSA.recover(msgHash, v, r, s);
-        require(recoveredAddr == arxAddress, "Invalid signature");
-        checkIns[arxAddress][msg.sender].push(timestamp);
-        if (checkIns[arxAddress][msg.sender].length == challenges[arxAddress].minimunCheckIns) {
-            challenges[arxAddress].totalStake -= challenges[arxAddress].perUserStake;
-            balances[msg.sender] += challenges[arxAddress].perUserStake;
-            succeedParticipants[arxAddress].push(msg.sender);
+        require(hasJoined[challengeId][msg.sender], "User has not joined the challenge");
+        bytes32 digest = getCheckInDigest(challengeId, timestamp, msg.sender);
+        require(!digestUsed[digest], "digest has been used");
+        require(challenges[challengeId].verifierAddress == ECDSA.recover(digest, v, r, s), "invalid signature");
+
+        digestUsed[digest] = true;
+        checkIns[challengeId][msg.sender].push(timestamp);
+        if (checkIns[challengeId][msg.sender].length == challenges[challengeId].minimunCheckIns) {
+            challenges[challengeId].totalStake -= challenges[challengeId].perUserStake;
```

Comment:
cache the variable `perUserStake` so you don't read from storage multiple times.
Also probably rename to stakePerUser